### ISSUE TITLE
TP gather self-copy ignores dst_offset when output device is not first shard

### ISSUE THREAD

**Issue by remichu-ai** [turboderp-org/exllamav3#233]
## Summary

There appears to be a bug in the native tensor-parallel gather CUDA kernel when the output device also contributes a non-zero-width shard and that output device is not the first shard in the concatenation order.

In `pg_gather_kernel`, the remote-copy path adds `dst_offset`, but the output-device/self-copy path computes `dst_offset` and then ignores it. As a result, the output device's local shard is copied to column offset 0 instead of its concatenation offset, corrupting gathered logits.

This also affects `TPBackendNCCL.gather()` because it currently delegates gather to the native fallback backend.

I checked current `origin/master` (`c5d9c65` at the time of writing), and the issue is still present:

```cpp
// exllamav3/exllamav3_ext/parallel/gather.cu
// self-copy path
size_t row = copy_t / ldim;
size_t col = copy_t % ldim;
size_t out_t = row * stride + col;  // missing + dst_offset
*((uint4*) (out_data_ptr + out_t)) = src[t];
```

The remote-copy path below it already does the right thing:

```cpp
size_t out_t = row * stride + col + dst_offset;
```

## Why this matters

For TP logits gather, if shards are concatenated as e.g. `[GPU0 shard, GPU1 shard]` and `output_device == GPU1`, GPU1's local lm_head output is written at offset 0 instead of after GPU0's shard. This can produce wildly incorrect logits/top-1 tokens even though each GPU's local lm_head result is correct.

The bug is hidden when:

- `world_size == 1` / gather returns early
- the output device is the first gather shard, so `dst_offset == 0`

## Minimal repro sketch

This uses the internal native TP backend directly and should fail on current master when `output_device` is the second device. It should pass after the one-line fix below.

```python
import multiprocessing as mp
import uuid

import torch

from exllamav3.model.model_tp_backend import TPBackendNative


def worker(device, active_devices, output_device, run_uuid, ldims, queue):
    torch.cuda.set_device(device)
    backend = TPBackendNative(
        device=device,
        active_devices=active_devices,
        output_device=output_device,
        init_method="",
        master=device == output_device,
        uuid=run_uuid,
    )

    try:
        local_idx = active_devices.index(device)
        ldim = ldims[local_idx]
        rows = 4

        # Unique constant per device makes column placement obvious.
        x = torch.full((rows, ldim), 100 + device, dtype=torch.float16, device=device)
        out = None
        if device == output_device:
            out = torch.empty((rows, sum(ldims)), dtype=torch.float16, device=device)

        backend.gather(x, out, active_devices, output_device, ldims)
        torch.cuda.synchronize(device)

        if device == output_device:
            queue.put(out.cpu())
    finally:
        backend.close()


def main():
    # Use any two visible CUDA devices. The important part is output_device is not
    # the first gather shard, so its dst_offset is non-zero.
    active_devices = [0, 1]
    output_device = 1
    ldims = [2, 3]
    run_uuid = "tp_gather_repro_" + uuid.uuid4().hex

    ctx = mp.get_context("spawn")
    queue = ctx.Queue()
    procs = [
        ctx.Process(target=worker, args=(device, active_devices, output_device, run_uuid, ldims, queue))
        for device in active_devices
    ]

    for proc in procs:
        proc.start()
    result = queue.get(timeout=30)
    for proc in procs:
        proc.join(timeout=30)
        assert proc.exitcode == 0, proc.exitcode

    expected = torch.tensor(
        [[100, 100, 101, 101, 101]] * 4,
        dtype=torch.float16,
    )
    print("result:", result)
    print("expected:", expected)
    assert torch.equal(result, expected)


if __name__ == "__main__":
    main()
```

## Model-level evidence

I saw this in a 2-GPU TP MiniMax-M3 run where the output device was GPU 4 and the first shard was on GPU 0:

- GPU 4 pre-gather lm_head shard matched the reference slice (`max_diff ~= 0.07`, fp16-level difference)
- full gathered logits were wrong (`max_diff ~= 18.68`), and top-1 changed
- a configuration where GPU 0 contributed only 640 lm_head channels produced an output token offset by exactly 640, consistent with GPU 4's shard being written at offset 0

After applying the one-line fix below and rebuilding:

- 2-GPU gathered logits matched the 1-GPU baseline within fp16 tolerance (`max_diff ~= 0.072`)
- first-token HF fixture checks passed for both native TP and NCCL TP gather fallback

## Proposed fix

```diff
diff --git a/exllamav3/exllamav3_ext/parallel/gather.cu b/exllamav3/exllamav3_ext/parallel/gather.cu
index 6a31868..0e38ac2 100644
--- a/exllamav3/exllamav3_ext/parallel/gather.cu
+++ b/exllamav3/exllamav3_ext/parallel/gather.cu
@@ -121,7 +121,7 @@ void pg_gather_kernel
                 {
                     size_t row = copy_t / ldim;
                     size_t col = copy_t % ldim;
-                    size_t out_t = row * stride + col;
+                    size_t out_t = row * stride + col + dst_offset;
                     *((uint4*) (out_data_ptr + out_t)) = src[t];
                 }
             }
```

Thanks!

**Comment by turboderp** [comment]
👍 Fixed in dev now.
